firebase-storage/src/main/java/com/google/firebase/storage/UploadTask.java

```diff
@@ -321,15 +337,18 @@ private boolean shouldContinue() {
     }
 
     boolean inErrorState = mServerException != null || mResultCode < 200 || mResultCode >= 300;
+    long deadLine = clock.elapsedRealtime() + this.maxSleepTime;
+    long currentTime = clock.elapsedRealtime() + this.sleepTime;
     // we attempt to recover by calling recoverStatus(true)
-    if (inErrorState && !recoverStatus(true)) {
-      // we failed to recover.
-      if (serverStateValid()) {
-        tryChangeState(INTERNAL_STATE_FAILURE, false);
+    if (inErrorState) {
+      if (currentTime > deadLine || !recoverStatus(true)) {
+        if (serverStateValid()) {
+          tryChangeState(INTERNAL_STATE_FAILURE, false);
+        }
+        return false;
       }
-      return false;
+      sleepTime = Math.max(sleepTime * 2, sleepTime + (minimumSleepInterval * 2));
```

Comment:
you probably want

`Math.max(sleepTime * 2, minimumSleepInterval)`

The way this is written, the first sleep time == 2000

Reply:
Oh wow. Nice catch! Will change